In [1]:
import pandas as pd
import numpy as np
import warnings

First the number of shots in the data set that have a value for the Shot column that is more than the Hole Score column

In [11]:
for year in range(2003,2017):
    with warnings.catch_warnings():
        warnings.simplefilter("ignore")
        data = pd.read_csv('./../data/rawdata/%d.txt' % year, sep = ';')
    data.columns = np.array([i.strip().replace(' ','_') for i in data.columns])
    tot_size = len(data)
    sus_size = float(len(data[data.Shot>data.Hole_Score]))
    pct = sus_size/tot_size*100
    print 'Year: %d' % year
    print 'Total Shots: %d' % tot_size
    print 'Suspect Shots: %d' % sus_size
    print 'Suspect Shots as percent of Total: %g' % pct + '\n'

Year: 2003
Total Shots: 1255867
Suspect Shots: 6744
Suspect Shots as percent of Total: 0.537

Year: 2004
Total Shots: 1272874
Suspect Shots: 7259
Suspect Shots as percent of Total: 0.570284

Year: 2005
Total Shots: 1279833
Suspect Shots: 7338
Suspect Shots as percent of Total: 0.573356

Year: 2006
Total Shots: 1298010
Suspect Shots: 7976
Suspect Shots as percent of Total: 0.614479

Year: 2007
Total Shots: 1211721
Suspect Shots: 7767
Suspect Shots as percent of Total: 0.640989

Year: 2008
Total Shots: 1206071
Suspect Shots: 8261
Suspect Shots as percent of Total: 0.684951

Year: 2009
Total Shots: 1137348
Suspect Shots: 7977
Suspect Shots as percent of Total: 0.701368

Year: 2010
Total Shots: 1160416
Suspect Shots: 8180
Suspect Shots as percent of Total: 0.70492

Year: 2011
Total Shots: 1125170
Suspect Shots: 8162
Suspect Shots as percent of Total: 0.725401

Year: 2012
Total Shots: 1102847
Suspect Shots: 8006
Suspect Shots as percent of Total: 0.725939

Year: 2013
Total Shots: 995304
Sus

I have manually checked 50 examples of the Hole Score field and verified the fact that it's correct with independent sources.

Ok, let's see a couple examples of the suspect shots to see if there is some normal pattern that emerges. I'll use the 2016 data.

In [102]:
with warnings.catch_warnings():
    warnings.simplefilter("ignore")
    data = pd.read_csv('./../data/rawdata/%d.txt' % 2016, sep = ';')
data.columns = np.array([i.strip().replace(' ','_') for i in data.columns])

In [103]:
id_cols = ['Permanent_Tournament_#','Year','Player_#','Round','Hole']

In [104]:
suspects = set(tuple(tup) for tup in data[(data.Shot>data.Hole_Score)][id_cols].values.tolist())

In [75]:
first_100_suspect_holes = [data[(data['Permanent_Tournament_#']==tourn) &
                                      (data.Year==year) & 
                                      (data['Player_#']==player) &
                                      (data.Round==round) &
                                      (data.Hole==hole)] 
                                 for u,(tourn,year,player,round,hole) in enumerate(suspects) if u<100]

First one

In [76]:
first_100_suspect_holes[0]

Tour_Code Tour_Description  Year  Tourn.#  Player_#  Course_#  \
94486         R         PGA TOUR  2016       60     27770       733   
94487         R         PGA TOUR  2016       60     27770       733   
94488         R         PGA TOUR  2016       60     27770       733   
94489         R         PGA TOUR  2016       60     27770       733   
94490         R         PGA TOUR  2016       60     27770       733   
94491         R         PGA TOUR  2016       60     27770       733   

       Permanent_Tournament_# Player_First_Name Player_Last_Name  Round  \
94486                     457            Camilo         Villegas      1   
94487                     457            Camilo         Villegas      1   
94488                     457            Camilo         Villegas      1   
94489                     457            Camilo         Villegas      1   
94490                     457            Camilo         Villegas      1   
94491                     457            Camilo         Villegas      1   

           ...       X_Coordinate  Y_Coordinate  Z_Coordinate  \
94486      ...              .0000         .0000             0   
94487      ...              .0000         .0000             0   
94488      ...              .0000         .0000             0   
94489      ...              .0000         .0000             0   
94490      ...              .0000         .0000             0   
94491      ...              .0000         .0000             0   

      Distance_from_Center  Distance_from_Edge        Date  Left/Right  \
94486                    0                   0  11/12/2015         NaN   
94487                    0                   0  11/12/2015         NaN   
94488                    0                   0  11/12/2015         NaN   
94489                    0                   0  11/12/2015         NaN   
94490                    0                   0  11/12/2015         NaN   
94491                    0                   0  11/12/2015         NaN   

      Strokes_Gained/Baseline  Strokes_Gained_Category Recovery_Shot  
94486                     0.0                      NaN            No  
94487                     0.0                      NaN            No  
94488                     0.0                      NaN            No  
94489                     0.0                      NaN            No  
94490                     0.0                      NaN            No  
94491                     0.0                      NaN            No  

[6 rows x 43 columns]

In [77]:
cols_of_interest = id_cols + ['Shot','Hole_Score','Distance','X_Coordinate','Y_Coordinate','Z_Coordinate','Distance_to_Hole_after_the_Shot']

In [78]:
first_100_suspect_holes[0][cols_of_interest]

Permanent_Tournament_#  Year  Player_#  Round  Hole  Shot Hole_Score  \
94486                     457  2016     27770      1     2     1          5   
94487                     457  2016     27770      1     2     2          5   
94488                     457  2016     27770      1     2     3          5   
94489                     457  2016     27770      1     2     4          5   
94490                     457  2016     27770      1     2     5          5   
94491                     457  2016     27770      1     2     6          5   

       Distance  X_Coordinate  Y_Coordinate Z_Coordinate  \
94486         0         .0000         .0000            0   
94487         0         .0000         .0000            0   
94488         0         .0000         .0000            0   
94489         0         .0000         .0000            0   
94490         0         .0000         .0000            0   
94491         0         .0000         .0000            0   

       Distance_to_Hole_after_the_Shot  
94486                                0  
94487                                0  
94488                                0  
94489                                0  
94490                                0  
94491                                0

This seems to be a completely useless player-hole unfortunately.

Let's see another example:

In [81]:
first_100_suspect_holes[1][cols_of_interest]

Permanent_Tournament_#  Year  Player_#  Round  Hole  Shot Hole_Score  \
1092196                     505  2016     25572      1    12     1          6   
1092197                     505  2016     25572      1    12     2          6   
1092198                     505  2016     25572      1    12     3          6   
1092199                     505  2016     25572      1    12     4          6   
1092200                     505  2016     25572      1    12     5          6   
1092201                     505  2016     25572      1    12     6          6   
1092202                     505  2016     25572      1    12     7          6   

         Distance  X_Coordinate  Y_Coordinate Z_Coordinate  \
1092196     10454    7,583.2410   10,040.9220      132.631   
1092197      5711    7,208.4300    9,747.9900      117.241   
1092198         0         .0000         .0000            0   
1092199      2752    7,400.0880    9,873.6100      126.864   
1092200      3464    7,189.8010    9,675.8110      125.716   
1092201       211    7,199.1090    9,690.7260      125.422   
1092202         6         .0000         .0000            0   

         Distance_to_Hole_after_the_Shot  
1092196                             6228  
1092197                              694  
1092198                              694  
1092199                             3251  
1092200                              217  
1092201                                6  
1092202                                0

Shot number 3 here is the extra one. Unfortunately, it is not always so clear cut as to which one is extra. For example:

In [82]:
first_100_suspect_holes[2][cols_of_interest]

Permanent_Tournament_#  Year  Player_#  Round  Hole  Shot Hole_Score  \
363430                      10  2016      6567      1    12     1          4   
363431                      10  2016      6567      1    12     2          4   
363432                      10  2016      6567      1    12     3          4   
363433                      10  2016      6567      1    12     4          4   
363434                      10  2016      6567      1    12     5          4   

        Distance  X_Coordinate  Y_Coordinate  Z_Coordinate  \
363430     10468    9,922.9750    9,091.1050       15.8870   
363431      5888   10,131.5740    8,647.0010       14.5510   
363432       176   10,121.6000    8,636.2400       15.1920   
363433       293   10,102.5000    8,651.4180       15.4470   
363434        93         .0000         .0000         .0000   

        Distance_to_Hole_after_the_Shot  
363430                             5632  
363431                              444  
363432                              382  
363433                               93  
363434                                0

This one contains coordinates for all shots that it should have coordinates for but it is not clear which one is the fake shot.

Unfortunately the data is full of example like this, with many of them very ambigious. I felt that it was a rabbit hole trying to create rules to identify the fake shots. Therefore I decided to drop all player-holes that had more shots recorded than the 'Hole Score' field says that the player took on the hole.

Additionally, there are player-holes in the data for which the number of shots matches the Hole Score field, but that contain shots that are not the last recorded shot on the hole but have 0 in one of the coordinate fields.

To show this, I will first drop the player-holes that contain more shots than the recorded Hole Score from the data in memory. Then I will run a simple query to show this phenomenon.

In [105]:
not_suspect = [True if tuple(tup) not in suspects else False for tup in data[id_cols].values.tolist()]

In [106]:
print 'Length data before drop: %d' % (len(data),)
print 'Number of not suspect shots: %d' % (sum(not_suspect),)
data = data.iloc[not_suspect]
print 'New length of data: %d' % (len(data),)

Length data before drop: 1141966
Number of not suspect shots: 1096101
New length of data: 1096101


First let's check if there are any player-holes with fewer shots in the data than the Hole Score field:

If we drop duplicates on columns that identify player-holes, the total length of the data - the number of shots - should equal the sum of the de-duped Hole-Score columns.

In [126]:
data = data[data['Permanent_Tournament_#']!=470] ## Dropping the match-play championship because there are blank Hole_Scores

In [128]:
len(data),data.drop_duplicates(subset=id_cols).Hole_Score.values.astype(int).sum(),len(data)-data.drop_duplicates(subset=id_cols).Hole_Score.values.astype(int).sum()

(1083861, 1083693, 168)

There seem to be 168 missing shots. Anyway, that is an aside. Now let's see if there are any shots that are not the last shot of a hole (Shot!=Hole_Score) that have 0 for a coordiante.

In [129]:
shots_with_zero_coords = data[(data.Shot!=data.Hole_Score) & ((data['X_Coordinate']==0) |
                                                              (data['Y_Coordinate']==0) | 
                                                              (data['Z_Coordinate']==0))]

In [130]:
len(shots_with_zero_coords)

69874

In [131]:
shots_with_zero_coords[cols_of_interest].head(15)

Permanent_Tournament_#  Year  Player_#  Round  Hole  Shot Hole_Score  \
186                       464  2016     10213      1    12     2          6   
1755                      464  2016     24140      1    14     3          4   
1876                      464  2016     24490      1    10     2          7   
2210                      464  2016     24846      1     5     2          6   
2390                      464  2016     25274      1    13     2          6   
3121                      464  2016     26331      1     1     2          6   
3124                      464  2016     26331      1     1     5          6   
4597                      464  2016     28132      1     5     2          6   
5110                      464  2016     29223      1     8     2          6   
6372                      464  2016     30944      1     3     2          7   
6383                      464  2016     30944      1     5     2          7   
12066                     464  2016     24361      2    12     2          6   
12553                     464  2016     24912      2     5     2          6   
13688                     464  2016     26881      2     3     2          7   
14357                     464  2016     27747      2     8     2          6   

       Distance  X_Coordinate  Y_Coordinate Z_Coordinate  \
186           0         .0000         .0000            0   
1755          0         .0000         .0000            0   
1876          0         .0000         .0000            0   
2210          0         .0000         .0000            0   
2390          0         .0000         .0000            0   
3121          0         .0000         .0000            0   
3124          5         .0000         .0000            0   
4597          0         .0000         .0000            0   
5110          0         .0000         .0000            0   
6372          0         .0000         .0000            0   
6383          0         .0000         .0000            0   
12066         0         .0000         .0000            0   
12553         0         .0000         .0000            0   
13688         0         .0000         .0000            0   
14357         0         .0000         .0000            0   

       Distance_to_Hole_after_the_Shot  
186                              14184  
1755                                36  
1876                             16884  
2210                             18972  
2390                             16452  
3121                              4772  
3124                                 0  
4597                             18972  
5110                             13140  
6372                             14904  
6383                             18972  
12066                            13752  
12553                            19224  
13688                            15768  
14357                            11844

Hence there are a lot of useless shots that have no coordinates recorded.

Let's see what percent of the data consists of holes with additional shots and/or holes with 0 in the coordinates before the last shot.

In [132]:
data.head()

Tour_Code Tour_Description  Year  Tourn.#  Player_#  Course_#  \
0         R         PGA TOUR  2016       10      2206       552   
1         R         PGA TOUR  2016       10      2206       552   
2         R         PGA TOUR  2016       10      2206       552   
3         R         PGA TOUR  2016       10      2206       552   
4         R         PGA TOUR  2016       10      2206       552   

   Permanent_Tournament_# Player_First_Name Player_Last_Name  Round  \
0                     464             David             Toms      1   
1                     464             David             Toms      1   
2                     464             David             Toms      1   
3                     464             David             Toms      1   
4                     464             David             Toms      1   

       ...       X_Coordinate  Y_Coordinate  Z_Coordinate  \
0      ...        10,662.7660    9,481.5800        80.447   
1      ...        10,795.3980    9,906.2420        86.468   
2      ...        10,842.1560    9,979.6790        91.197   
3      ...              .0000         .0000             0   
4      ...        10,203.1030   10,519.6630       100.531   

  Distance_from_Center  Distance_from_Edge        Date  Left/Right  \
0                  541                  56  10/15/2015           R   
1                  103                 203  10/15/2015         NaN   
2                   47                 602  10/15/2015           L   
3                    0                   0  10/15/2015         NaN   
4                   27                 550  10/15/2015           R   

  Strokes_Gained/Baseline  Strokes_Gained_Category Recovery_Shot  
0                  -0.179              Off the Tee            No  
1                  -0.217    Approach to the Green            No  
2                   0.372         Around the Green            No  
3                   0.113                  Putting            No  
4                   0.462    Approach to the Green            No  

[5 rows x 43 columns]

In [163]:
for year in range(2003,2017):
    data = pd.read_csv('./../data/rawdata/%d.txt' % year, sep = ';') ## read in data

    ## making columns usable
    data.columns = [i.strip().replace(' ','_') for i in data.columns]

    ## converting all coords to str to strip out white space
    data['X_Coordinate'] = [str(i).replace(' ','') for i in data['X_Coordinate']]
    data['Y_Coordinate'] = [str(i).replace(' ','') for i in data['Y_Coordinate']]
    data['Z_Coordinate'] = [str(i).replace(' ','') for i in data['Z_Coordinate']]
    
    ## getting rid of commas
    data['X_Coordinate'] = [str(i).replace(',','') for i in data['X_Coordinate']] 
    data['Y_Coordinate'] = [str(i).replace(',','') for i in data['Y_Coordinate']]
    data['Z_Coordinate'] = [str(i).replace(',','') for i in data['Z_Coordinate']]
    
    ## placing negative signs in proper place
    data['X_Coordinate'] = ['-' + str(i)[:len(str(i))-1] if str(i)[len(str(i))-1]=='-' else i for i in data['X_Coordinate']] 
    data['Y_Coordinate'] = ['-' + str(i)[:len(str(i))-1] if str(i)[len(str(i))-1]=='-' else i for i in data['Y_Coordinate']]
    data['Z_Coordinate'] = ['-' + str(i)[:len(str(i))-1] if str(i)[len(str(i))-1]=='-' else i for i in data['Z_Coordinate']]
    
    ## converting to number
    data['X_Coordinate'] = pd.to_numeric(data['X_Coordinate'])
    data['Y_Coordinate'] = pd.to_numeric(data['Y_Coordinate'])
    data['Z_Coordinate'] = pd.to_numeric(data['Z_Coordinate'])

    ## Dropping the match-play championship because there are blank Hole_Scores
    data = data[data['Permanent_Tournament_#']!=470]
    
    ## Ensuring Hole Score and Shot are numerics
    data.Hole_Score = pd.to_numeric(data.Hole_Score)
    data.Shot = pd.to_numeric(data.Shot)

    len_before = len(data)    
    id_cols = ['Year','Permanent_Tournament_#','Player_#','Round','Hole'] ## cols to identify player-hole
    
    ## player-holes that have more shots than the Hole Score
    to_remove = set([tuple(i) for i in data[data.Shot>data.Hole_Score][id_cols].values.tolist()])
    
    ## player-holes that have coordinates that are 0 for shots that are not the last of the hole
    to_remove.update([tuple(i) for i in 
                      data[(data.Shot!=data.Hole_Score) & ((data.X_Coordinate==0) | 
                                                           (data.Y_Coordinate==0) |
                                                           (data.Z_Coordinate==0))][id_cols].values.tolist()])
    
    ## find number of shots that are alright on both conditions
    after =  sum(1 for i in data[id_cols].values.tolist() if tuple(i) not in to_remove)
    pct_lost = (before - after)/float(before)*100
    print 'Year: %d . Data shrunk by %g percent.' % (year,pct_lost)
    print 'Number of fine shots: %d ' % after + '\n'

Year: 2003 . Data shrunk by 15.5201 percent.
Number of fine shots: 1060955 

Year: 2004 . Data shrunk by 11.5745 percent.
Number of fine shots: 1110507 

Year: 2005 . Data shrunk by 9.9165 percent.
Number of fine shots: 1131329 

Year: 2006 . Data shrunk by 10.5709 percent.
Number of fine shots: 1123111 

Year: 2007 . Data shrunk by 14.987 percent.
Number of fine shots: 1067650 

Year: 2008 . Data shrunk by 14.0376 percent.
Number of fine shots: 1079574 

Year: 2009 . Data shrunk by 21.9182 percent.
Number of fine shots: 980604 

Year: 2010 . Data shrunk by 19.8113 percent.
Number of fine shots: 1007063 

Year: 2011 . Data shrunk by 22.522 percent.
Number of fine shots: 973021 

Year: 2012 . Data shrunk by 23.8192 percent.
Number of fine shots: 956730 

Year: 2013 . Data shrunk by 31.0133 percent.
Number of fine shots: 866381 

Year: 2014 . Data shrunk by 24.7353 percent.
Number of fine shots: 945224 

Year: 2015 . Data shrunk by 21.9431 percent.
Number of fine shots: 980291 

Year: 20